### ISSUE TITLE
Argument passed to verify() is of type MyClassToSpy $$SpringCGLIB$$0 and is not a mock!

### ISSUE THREAD

**Issue by spyro2000** [antoinemeyer/mock-in-bean#23]
Version: boot2-v1.5.2
Spring Version: 3.1.2

Hi there,

I tried to replace a `@SpyBean` (which dirties my Spring context) with this setup:

```java
 @SpyInBean(MyCallingClass.class)
 private MyClassToSpy classToSpy;
````
The test itself runs fine but I can't verify the call with Mockito:

```java
 verify(classToSpy).doSomething();
````

This returns `Argument passed to verify() is of type MyClassToSpy $$SpringCGLIB$$0 and is not a mock!`

What did I do wrong here? Is Spring Boot 3 just not supported?

Thank you!

**Comment by antoinemeyer** [comment]
Hello,

Spring boot 3 should work with no problem.
Did this issue happen when migrating to spring boot 3?

Is there any way you could isolate the issue in a small snippet for me to reproduce?

**Comment by spyro2000** [comment]
Hi @antoinemeyer,

thanks for your quick response :)
I can't tell whether it's actually a problem with Spring 3, my project was already created with it.

Unfortunately, the project is very complex, so maybe it's a problem with my setup. I am within an `@SpringBootTest` and just replaced my `@SpyBean` with `@SpyInBean` and specified the class, with I expect to do the actual call I want to spy as an argument. 

I am a little bit confused, if `$$SpringCGLIB$$0` is actually the expected class?

I hope I will find the time to reproduce this in a smaller project in the upcoming days.

**Comment by antoinemeyer** [comment]
Could you try to use version `boot2-v2.1-RC` and see if you face the same issue?

**Comment by spyro2000** [comment]
Hi @antoinemeyer,

ok, did this but got an error when starting the test:

```
{"level":"WARN ", "class":"TestContextManager", "log_message":"Caught exception while invoking 'beforeTestMethod' callback on TestExecutionListener [com.teketik.test.mockinbean.MockInBeanTestExecutionListener] for test method [public void com.bmw.cc.tsdiagnosiswidget.api.app.v1.shared.boundary.CallbackResourceITest.shouldPublishMessageViaWebSocketIfAuthorized()] and test instance [com.bmw.cc.tsdiagnosiswidget.api.app.v1.shared.boundary.CallbackResourceITest@1f33e4b5]"}
java.lang.IllegalArgumentException: Superclass has no null constructors but no arguments were given
	at org.springframework.cglib.proxy.Enhancer.emitConstructors(Enhancer.java:1025) ~[spring-core-6.0.11.jar:6.0.11]
```

Did replace

```java
@SpyBean
private ConnectionBF publisher;
````

with

```java
@SpyInBean(CallbackBF.class)
private ConnectionBF publisher;
````

**Comment by antoinemeyer** [comment]
Thank you for sharing those details.
Without a test case to reproduce, it's really hard to find where the issue could be ...
I have some ideas and will try to poke around but please let me know if you manage to isolate the issue, that would help dramatically.

**Comment by spyro2000** [comment]
Yes, of course I will do. Thank you in advance!

**Comment by inkassso** [comment]
@antoinemeyer I have reproduced the issue with a minimal working sample, please see my repo [inkassso/mock-in-bean-issue-23](https://github.com/inkassso/mock-in-bean-issue-23).

What I have found, the issue occurs when `@SpyInBean` is applied on a service proxy. Spring creates a proxy for components, where method invocation from another component has to be intercepted. The mechanism is of course widely used for transaction management, async execution, AOP (as in my case) etc.

The issue with this is, that when setting up a mock, Mockito seems to be unaware of any Spring proxies (and frankly why should it?) and when storing information about a mock internally in a Map, the hashCode of the proxy is used. Quite suprisingly, during the verification process, Mockito actually is aware of Spring proxies. When checking, if the provided instance is actually a mock, it is looking up the mock information in said Map, for which the hashCode of the mock (the argument passed to `verify`) is computed. However, this time it unwraps the Spring proxy beforehand (even through the outer spy), because it's expecting the spy to be within the proxy, not outside. Of course, the hashCode of actual service instance is different than the one of the proxy, under which the information is actually stored. That for mockito means, that the provided argument is not a spy, even though it clearly is.

Please take a look at the testcases. I also included a test with a workaround, which removes the proxy (and any interceptors with it).

Also, `@MockInBean` has no issue, because in that case Mockito only considers the resolvable type of the declared test field, without reading it from the actual service instance like in case of the spy.

**Comment by spyro2000** [comment]
Thank you @antoinemeyer, this was a very impressive piece of reverse-engineering.

**Comment by inkassso** [comment]
Thank you, @spyro2000.

As to why Mockito unwraps the proxy, it's a customization by Spring only applied in the verification step, see [SpringBootMockResolver](https://github.com/spring-projects/spring-boot/blob/v3.3.5/spring-boot-project/spring-boot-test/src/main/java/org/springframework/boot/test/mock/mockito/SpringBootMockResolver.java). Regarding mock creation and why that isn't necessary with `@SpyBean`, it is Spring providing Mockito with the actual service instance instead of the proxy, so Mockito doesn't have to unwrap anything. Spring creates the spy during startup when the bean is created and before wrapping the service (or the spy) in the proxy. I suppose that cannot be done from this library, given the context is already constructed when the Test Execution Listener is invoked.

However, the proxy target can be set using reflection, just like this library injects the spy into the target service. I added another testcase (nr. 4) where the proxy is unwrapped and instead of injecting the spy directly into the target service, it is injected in the wrapping proxy. The target service keeps the reference to the proxy and when the proxy is invoked, it invokes the spy, which then invokes the actual service instance. Works like a charm, and while it isn't as elegant, eventually the service vs. spy structure is the same as with Spring's `@SpyBean`.

**Comment by antoinemeyer** [comment]
Thanks @inkassso for providing the analysis and a reproducible test case.

A solution would be to distinguish the original value from the mockable value which would be that proxy containing the ultimate target.
The tests would leverage the mockable object that would be injected in the beans and the original value would be reinjected in the beans after tests.

I opened https://github.com/antoinemeyer/mock-in-bean/pull/28 with suggested changes. I added a test with the test case you provided. Does that fix your issue?
